File: algorithms/0027/main.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
  // left 指向第一个值为 val 的 index
  // right 指向最后一个值不为 val 的 index
  let left = 0
  let right = nums.length - 1
  while (true) {
    while (left < nums.length && nums[left] !== val) {
      left++
    }
    while (right >= 0 && nums[right] === val) {
      right--
    }
    if (left >= right) {
      break
    }
    const tmp = nums[left]
    nums[left] = nums[right]
    nums[right] = tmp
  }
  return left
}

module.exports = removeElement

File: algorithms/0027/main.go
package p0027

func removeElement(nums []int, val int) int {
	left, right := 0, len(nums)-1
	for {
		for left < len(nums) && nums[left] != val {
			left++
		}
		for right >= 0 && nums[right] == val {
			right--
		}
		if left >= right {
			break
		}
		nums[left], nums[right] = nums[right], nums[left]
	}
	return left
}
